core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -768,13 +692,151 @@ private void sendFailShard(ShardRouting shardRouting, String message, @Nullable
     private class FailedShardHandler implements Callback<IndexShard.ShardFailure> {
         @Override
         public void handle(final IndexShard.ShardFailure shardFailure) {
-            final IndexService indexService = indicesService.indexService(shardFailure.routing.shardId().getIndex());
             final ShardRouting shardRouting = shardFailure.routing;
             threadPool.generic().execute(() -> {
                 synchronized (mutex) {
-                    failAndRemoveShard(shardRouting, indexService, true, "shard failure, reason [" + shardFailure.reason + "]", shardFailure.cause);
+                    failAndRemoveShard(shardRouting, true, "shard failure, reason [" + shardFailure.reason + "]", shardFailure.cause);
                 }
             });
         }
     }
+
+    public interface Shard {
+
+        /**
+         * Returns the shard id of this shard.
+         */
+        ShardId shardId();
```

Comment:
this method and the others in this interface are used in tons of places. I leave this to a follow-up PR (this one is already getting bigger and bigger).